pkg/token/filecache.go

```diff
@@ -0,0 +1,277 @@
+		_, _ = fmt.Fprintf(os.Stderr, "Cache file %s does not exist.\n", filename)
+	}
+
+	return FileCacheProvider{
+		creds,
+		clusterID,
+		profile,
+		roleARN,
+		cachedToken,
+	}, nil
+}
+
+// Retrieve() implements the Provider interface, returning the cached token if is not expired,
+// otherwise fetching the token from the underlying Provider and caching the results on disk
+// with an expiration time.
+func (f *FileCacheProvider) Retrieve() (credentials.Value, error) {
+	if !f.cachedToken.IsExpired() {
+		// use the cached token
+		return f.cachedToken.Token, nil
+	} else {
+		_, _ = fmt.Fprintf(os.Stderr, "No cached token available.  Refreshing...\n")
+		// fetch the token from the underlying Provider
+		token, err := f.credentials.Get()
+		if err == nil {
+			if expiration, err := f.credentials.ExpiresAt(); err == nil {
+				// underlying provider supports Expirer interface, so we can cache
+				filename := CacheFilename()
+				// do file locking on cache to prevent inconsistent writes
+				lock := newFlock(filename)
+				defer lock.Unlock()
+				// wait up to a second for the file to lock
+				ctx, cancel := context.WithTimeout(context.TODO(), time.Second)
+				defer cancel()
+				ok, err := lock.TryLockContext(ctx, 250 * time.Millisecond) // try to lock every 1/4 second
+				if ok {
+					f.cachedToken = cacheToken{
+						token,
+						expiration,
+						nil,
+					}
+					// don't really care about read error.  Either read the cache, or we create a new cache.
+					cache, _ := readCacheWhileLocked(filename)
+					if _, ok := cache.ClusterMap[f.clusterID]; !ok {
+						// first use of this cluster id
+						cache.ClusterMap[f.clusterID] = map[string]map[string]cacheToken{}
+					}
+					if _, ok := cache.ClusterMap[f.clusterID][f.profile]; !ok {
+						// first use of this profile
+						cache.ClusterMap[f.clusterID][f.profile] = map[string]cacheToken{}
+					}
+					cache.ClusterMap[f.clusterID][f.profile][f.roleARN] = f.cachedToken
+					err = writeCacheWhileLocked(filename, cache)
+					if err != nil {
+						// can't write cache, but still return the credential
+						_, _ = fmt.Fprintf(os.Stderr, "Unable to update token cache %s: %v\n", filename, err)
+					} else {
+						_, _ = fmt.Fprintf(os.Stderr, "Updated cached token\n")
+					}
+				} else {
+					// can't get write lock to create/update cache, but still return the credential
+					_, _ = fmt.Fprintf(os.Stderr, "Unable to write lock file %s: %v\n", filename, err)
+				}
+			} else {
+				// credential doesn't support expiration time, so can't cache, but still return the credential
+				_, _ = fmt.Fprintf(os.Stderr, "Unable to cache token: %v\n", err)
+			}
+		}
+		return token, err
+	}
+}
+
+// IsExpired() implements the Provider interface, deferring to the cached token first,
+// but fall back to the underlying Provider if it is expired.
+func (f *FileCacheProvider) IsExpired() bool {
+	return f.cachedToken.IsExpired() && f.credentials.IsExpired()
```

Comment:
Any chance that `cachedToken` or `credentials` might be empty here?

Reply:
'cachedToken' is an object.  'credentials' is a pointer, but is passed in to NewFileCacheProvider().  Conceivably, I suppose the user could pass in nil for that.  I will add a check for that in NewFileCacheProvider() and return an error if nil.